crates/ra_assists/src/fill_struct_fields.rs

```diff
@@ -1,51 +1,99 @@
 use std::fmt::Write;
 
-use hir::{AdtDef, Ty, source_binder};
-use hir::db::HirDatabase;
+use hir::{AdtDef, Ty, db::HirDatabase, source_binder::function_from_child_node, Name, StructField,};
 
 use ra_syntax::ast::{self, AstNode};
 
 use crate::{AssistCtx, Assist, AssistId};
 
-pub(crate) fn fill_struct_fields(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
-    let struct_lit = ctx.node_at_offset::<ast::StructLit>()?;
-    let named_field_list = struct_lit.named_field_list()?;
+struct FillStructFields<'a, 'b: 'a, DB> {
+    borrowed_ctx: &'a mut AssistCtx<'b, DB>,
+    named_field_list: &'a ast::NamedFieldList,
+    struct_fields: Vec<StructField>,
+    struct_lit: &'a ast::StructLit,
+}
```

Comment:
Do we really need this struct though? Can we just turn its methods into functions, which accept a bunch of arguments?

Reply:
I did it first time but ended up with a mess of duplicated function signatures